### PR TITLE
feat(dynamics365): add Dataverse and Business Central tasks

### LINKED ISSUE #61: Microsoft Dynamics 365
## Description
Create a new sub plugin for **Microsoft Dynamics 365** to enable orchestration of CRM and ERP workflows. This plugin will allow Kestra to interact with the **Dataverse Web API** (for Sales, Marketing, and Service) and the **Business Central API v2.0** (for ERP/Finance).

This plugin fits into the **BUSINESS** category and serves as an enterprise-grade alternative to Odoo, Pipedrive, and SAP integrations.

## Potential Sub-packages
Since Dynamics 365 is a suite of distinct products, we propose organizing the plugin into the following sub-packages:

1. **`io.kestra.plugin.microsoft.dynamics365.dataverse`** (CRM / Data Platform)
   * Focus: Sales, Customer Service, and custom Dataverse tables.
   * Tasks: `Query`, `Upsert`, `Delete`, `Trigger`.
2. **`io.kestra.plugin.microsoft.dynamics365.businesscentral`** (ERP for SMB)
   * Focus: Finance, Inventory, and Purchasing for small to medium businesses.
   * Tasks: `GetInvoice`, `CreateCustomer`, `SyncInventory`.
3. **`io.kestra.plugin.microsoft.dynamics365.finance`** (Enterprise ERP)
   * Focus: Large scale finance and supply chain operations via the Data Management Framework (DMF).

---

### API Reference & Documentation
There is no official modern Java SDK; the integration should be built against the **OData v4 REST APIs**.

* **Business Central API v2.0:** * [Official API Reference](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/api-reference/v2.0/)
  * **Base URL:** `https://api.businesscentral.dynamics.com/v2.0/{tenantId}/{environment}/api/v2.0`
* **Dataverse (CRM) Web API:**
  * [Official API Reference](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/overview)
  * **Base URL:** `https://{orgName}.api.crm.dynamics.com/api/data/v9.2/`

### Important Routes (Business Central v2.0)
| Resource | Route | Description |
| :--- | :--- | :--- |
| **Companies** | `/companies` | [Link](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/api-reference/v2.0/resources/dynamics_company) - Entry point for most requests. |
| **Customers** | `/companies({id})/customers` | [Link](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/api-reference/v2.0/resources/dynamics_customer) - Manage CRM entities in ERP. |
| **Sales Invoices** | `/companies({id})/salesInvoices` | [Link](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/api-reference/v2.0/resources/dynamics_salesinvoice) - Financial document management. |
| **Items** | `/companies({id})/items` | [Link](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/api-reference/v2.0/resources/dynamics_item) - Inventory and catalog sync. |

### Authentication
The plugin should utilize **Microsoft Entra ID (Azure AD)** via:
* **OAuth2 Client Credentials Flow** (Service-to-Service).
* **Connection Strings/Tokens** consistent with the existing `plugin-microsoft365`.

## Use Cases
* **Automated Billing:** Automatically generate a Sales Invoice in Business Central when a Kestra flow completes a billable data processing job.
* **Lead Orchestration:** Syncing high-value leads from external databases or webhooks directly into Dynamics 365 Sales (Dataverse).
* **Inventory Alerts:** Triggering Slack/Email notifications via Kestra when Business Central item levels fall below a certain threshold.

## Category
**BUSINESS**

### PR DESCRIPTION
**Pull request by fdelbrayelle**
## Summary

- Adds `io.kestra.plugin.microsoft365.dynamics365.dataverse` sub-package with `Query`, `Upsert`, and `Delete` tasks for the Dataverse Web API v9.2
- Adds `io.kestra.plugin.microsoft365.dynamics365.businesscentral` sub-package with `ListCompanies`, `GetCustomer`, `CreateCustomer`, `GetInvoice`, `CreateInvoice`, and `ListItems` tasks for Business Central API v2.0
- Reuses the existing `MicrosoftGraphConnectionInterface` auth pattern (tenantId + clientId + clientSecret via Azure AD OAuth2 Client Credentials flow)
- Includes WireMock-based functional tests for all 9 tasks, SVG icons, metadata YAML, and `package-info.java` files

closes: https://github.com/kestra-io/plugin-microsoft365/issues/61

## Test plan

- [x] All 72 unit tests pass (`./gradlew test`)
- [x] Plugin builds cleanly (`./gradlew shadowJar`)
- [x] Kestra EE: both `Dynamics 365 Dataverse` and `Dynamics 365 Business Central` sub-groups registered in the Plugins page
- [x] Kestra EE: `dataverse.Query` task executes and fails at expected secret resolution step (no NPE, no plugin load error)
- [x] Kestra EE: `businesscentral.ListCompanies` task executes and fails at expected secret resolution step
- [x] Non-regression: existing `oneshare.List` task unaffected

### CODE REVIEW

File: src/test/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/QueryTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Map;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+@KestraTest
+class QueryTest {
+
+    @Inject
+    private RunContextFactory runContextFactory;
```
Comment (fdelbrayelle) [inline src/test/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/QueryTest.java:29]: BLOCK [TESTS] No happy-path functional test — run() is never called

Problem: All 9 new test classes only exercise the Lombok builder and the Output POJO. RunContextFactory is injected but never used. The run() method is not invoked in any test, so pagination, URL construction, OData parsing, error handling, and metric emission are all untested.

The coding standards are explicit: every task must include at least one functional test covering the happy path. Testing only edge cases or failure scenarios is not sufficient.

Fix: Add WireMock-based functional tests. Stub the Azure token endpoint and the OData/BC endpoints. Call task.run(runContext) and assert on the returned Output. Use: testImplementation "org.wiremock:wiremock-jetty12"

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/AbstractDynamics365Task.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * Acquires a Bearer access token for the given OAuth2 scope using the configured credentials.
+     * A fresh token is obtained on every call — no caching between runs.
+     */
+    protected String getAccessToken(RunContext runContext, String scope) throws IllegalVariableEvaluationException {
+        var rTenantId = runContext.render(this.tenantId).as(String.class).orElseThrow(
+            () -> new IllegalArgumentException("tenantId is required for authentication"));
+        var rClientId = runContext.render(this.clientId).as(String.class).orElseThrow(
+            () -> new IllegalArgumentException("clientId is required for authentication"));
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/AbstractDynamics365Task.java:86]: BLOCK [SECURITY / CORRECTNESS] Both auth credentials set: clientSecret silently wins

Problem: When a user supplies both clientSecret and pemCertificate the method silently uses clientSecret. The Schema description for both properties says "Either clientSecret OR pemCertificate must be provided, not both." Silent preference violates the documented contract and can mask configuration mistakes.

Fix: Add a guard at the top of getAccessToken():
if (StringUtils.isNotBlank(rClientSecret) && StringUtils.isNotBlank(rPemCertificate)) {
    throw new IllegalArgumentException("Provide either clientSecret or pemCertificate, not both");
}

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        urlBuilder.append("?$top=").append(rTop);
+        if (rFilter != null) {
+            urlBuilder.append("&$filter=").append(URLEncoder.encode(rFilter, StandardCharsets.UTF_8));
+        }
+        if (rSelect != null) {
+            urlBuilder.append("&$select=").append(URLEncoder.encode(rSelect, StandardCharsets.UTF_8));
+        }
+
+        List<Map<String, Object>> allRecords = new ArrayList<>();
+        String nextUrl = urlBuilder.toString();
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java:179]: REQUEST_CHANGES [CORRECTNESS] FETCH silently drops all pages after the first

Problem: When fetchType is FETCH the loop exits after the first page regardless of @odata.nextLink. With the default top=100 any result set larger than 100 records is silently truncated — the user has no indication records were dropped.

The Output schema says "List of entity records returned; populated when fetchType is FETCH" with no truncation warning, which makes this a silent data-loss bug in realistic scenarios.

Fix: Either (a) follow pagination for FETCH too (capping only via the top property), or (b) document the truncation clearly in the Schema description and log a warning at WARN level when nextLink is present but discarded:
if (page.getOdataNextLink() != null) {
    logger.warn("FETCH mode: additional pages exist but will not be retrieved. Use STORE to fetch all records.");
}

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+            """
+    )
+    @NotNull
+    @Builder.Default
+    @PluginProperty(group = "processing")
+    private Property<FetchType> fetchType = Property.ofValue(FetchType.FETCH);
+
+    private static final ObjectMapper MAPPER = JacksonMapper.ofJson();
+
+    @Override
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java:151]: REQUEST_CHANGES [CORRECTNESS] fetchType=FETCH_ONE is silently handled as FETCH

Problem: The FetchType enum has three values (FETCH_ONE, FETCH, STORE). Only STORE is explicitly branched on; FETCH_ONE falls through to the same path as FETCH, returning up to top records when only one was requested. The Kestra standard requires support for all three FetchType values.

Fix: Add explicit FETCH_ONE handling. After the first page is retrieved, take only the first record from page.getValue() and return it in the output. Update the Output to include a single-record field for this case.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    )
+    @PluginProperty(group = "processing")
+    private Property<String> select;
+
+    @Schema(
+        title = "Maximum records per page ($top)",
+        description = """
+            Maximum number of records to return per OData page.
+            When `fetchType` is `STORE`, all pages are followed regardless of this value.
+            Defaults to 100.
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java:131]: REQUEST_CHANGES [DRY] parseAndThrowODataError duplicated across Query, Upsert, and Delete

Problem: The method is copy-pasted in all three Dataverse task classes. Any fix or improvement must be applied in three places.

Fix: Move parseAndThrowODataError to AbstractDataverseTask as a protected static method. Remove all three private copies.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/businesscentral/ListItems.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    @Override
+    public Output run(RunContext runContext) throws Exception {
+        var logger = runContext.logger();
+
+        var rCompanyId = runContext.render(companyId).as(String.class).orElseThrow();
+        var rFilter = runContext.render(filter).as(String.class).orElse(null);
+        var rTop = runContext.render(top).as(Integer.class).orElse(100);
+
+        var token = getAccessToken(runContext, scope());
+        var urlBuilder = new StringBuilder(baseUrl(runContext))
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/businesscentral/ListItems.java:105]: REQUEST_CHANGES [DRY / COUPLING] Cross-class static call to ListCompanies.parseAndThrowError()

Problem: ListItems, GetCustomer, GetInvoice, CreateCustomer, and CreateInvoice all call ListCompanies.parseAndThrowError(). This creates tight coupling between unrelated task classes and makes ListCompanies impossible to refactor without hunting all dependents.

Fix: Move parseAndThrowError to AbstractBusinessCentralTask as a protected static method (same fix as for the Dataverse side in AbstractDataverseTask). All BC task classes then call the inherited method directly.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Upsert.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                .addHeader("If-Match", "*")
+                .uri(URI.create(url))
+                .method("PATCH")
+                .body(HttpRequest.StringRequestBody.builder().content(body).build())
+                .build();
+
+            var response = client.request(request, String.class);
+            var statusCode = response.getStatus().getCode();
+
+            // 204 No Content on success (upsert returns no body)
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Upsert.java]: NIT [STYLE] Comment restates what the HTTP spec already says

Problem: "// 204 No Content on success (upsert returns no body)" describes the standard HTTP 204 semantics — the code does not need to explain this. The coding standards forbid comments that restate what the code already says.

Fix: Remove this comment line.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
+package io.kestra.plugin.microsoft365.dynamics365.dataverse;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.core.type.TypeReference;
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Query.java]: NIT [STYLE] Unused import: com.fasterxml.jackson.core.type.TypeReference

Problem: TypeReference is imported but Query only deserializes via MAPPER.readValue(body, ODataResponse.class) — it never uses TypeReference. Unused imports must be removed per the coding standards.

Fix: Delete the import line.

**Comment by fdelbrayelle** [comment]
## Kestra Plugin Code Review — commit c5b892e

### Kestra Guidelines — 7 findings

**BLOCK**
- QueryTest / UpsertTest / DeleteTest and all 6 BC tests: No run() call in any of the 9 new test classes. RunContextFactory is injected but never used. Builder-only tests do not satisfy the mandatory happy-path requirement. Must add WireMock stubs for the Azure token endpoint and the OData/BC endpoints and call task.run(runContext) with output assertions. Dependency: testImplementation "org.wiremock:wiremock-jetty12".

**REQUEST CHANGES**
- AbstractDynamics365Task#getAccessToken: When both clientSecret and pemCertificate are provided, clientSecret silently wins. The documented contract says mutual exclusion. A guard must throw IllegalArgumentException instead of silently preferring one.
- Query#run (FETCH mode): fetchType=FETCH stops at the first page regardless of @odata.nextLink. With default top=100, any result set >100 records is silently truncated. Either follow pagination or emit a WARN and document the truncation in @Schema.
- Query#run (FETCH_ONE): FetchType.FETCH_ONE is silently treated as FETCH. Kestra standards require all three fetch modes to be explicitly handled.
- Query / Upsert / Delete: parseAndThrowODataError is copy-pasted across all three Dataverse tasks. Move it to AbstractDataverseTask as a protected static method.
- ListItems / GetCustomer / GetInvoice / CreateCustomer / CreateInvoice all call ListCompanies.parseAndThrowError(). Move to AbstractBusinessCentralTask.

**SUGGESTION**
- Metadata filenames (dynamics365-businesscentral.yaml, dynamics365-dataverse.yaml) are inconsistent with the flat leaf-name convention used by sibling files (oneshare.yaml, teams.yaml). Confirm and align.
- index.yaml not updated. Body still mentions only Microsoft Graph API; Dynamics 365 support warrants a description update.

**NIT**
- Comments on lines 117 of Upsert.java and Delete.java (204 No Content) restate HTTP spec; remove per no-restatement rule.
- Query.java line 4: import com.fasterxml.jackson.core.type.TypeReference is unused; remove it.

---

### Security (OWASP Top 10) — 1 finding

**REQUEST CHANGES (A02)**
- AbstractDynamics365Task#getAccessToken: When both credentials are rendered from Pebble templates, verify that no exception message path embeds the raw secret values. Currently safe, but warrant an explicit review before merge.

---

### Business Requirements — met (partial)

Scope is correct. All 9 described tasks are present. The gap is test coverage, not scope.

---

**Verdict: BLOCK**

The blocking issue is the complete absence of happy-path functional tests across all 9 new tasks. Additional REQUEST_CHANGES items (auth ambiguity, FETCH truncation, FETCH_ONE unhandled, DRY violations) must also be resolved before merge.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/AbstractDynamics365Task.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            """
+    )
+    @PluginProperty(group = "advanced")
+    protected Property<String> pemCertificate;
+
+    /**
+     * Acquires a Bearer access token for the given OAuth2 scope using the configured credentials.
+     * A fresh token is obtained on every call — no caching between runs.
+     */
+    public String getAccessToken(RunContext runContext, String scope) throws IllegalVariableEvaluationException {
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/AbstractDynamics365Task.java]: test body

Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/AbstractDynamics365Task.java]: 🟠 [Guidelines] getAccessToken widened to public to enable Mockito spy — production changed for test convenience

Problem: This method was protected, intentionally scoped to subclasses. It was made public solely so Mockito spy calls from test classes in sub-packages compile. The guidelines state explicitly: "Do not write production code specifically to satisfy tests."

Fix: Revert to protected. Use a package-private or test-only base class in src/test that overrides getAccessToken to return a fixed value, which each test class extends. This removes the need to widen the production visibility contract.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/businesscentral/AbstractBusinessCentralTask.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    protected Property<String> environment;
+
+    @Schema(
+        title = "Business Central API endpoint",
+        description = """
+            Override the Business Central API base URL.
+            Useful for sovereign clouds or for testing. Defaults to `https://api.businesscentral.dynamics.com` when not set.
+            """
+    )
+    @PluginProperty(group = "advanced", hidden = true)
```
Comment (fdelbrayelle) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/businesscentral/AbstractBusinessCentralTask.java:42]: 🟡 [Guidelines/YAGNI] apiEndpoint added primarily to redirect BC tests to WireMock

Problem: The sovereign-cloud justification is plausible, but the method signature change (scope() now requires RunContext, propagated to all 6 BC task classes) was clearly driven by test redirection. Compare to Dataverse where orgUrl already plays this role cleanly, with no base-class surgery.

Fix: If sovereign-cloud support is a real requirement, keep the property but remove the "or for testing" language from the description. If it is test-only, remove it and redirect tests via a test-only subclass that overrides baseUrl() and scope() directly.

**Comment by fdelbrayelle** [comment]
## Kestra Plugin Code Review — fix(dynamics365): address code review findings (18914f0)

### Prior Issue Verification

All 8 prior findings are resolved:
1. Missing functional tests — RESOLVED: all 9 test classes now wire WireMock stubs and call task.run()
2. Silent auth ambiguity — RESOLVED: early IllegalArgumentException when both clientSecret and pemCertificate are set
3. FETCH pagination — RESOLVED: Query follows @odata.nextLink for FETCH and STORE
4. FETCH_ONE unhandled — RESOLVED: forces top=1, breaks after first page, returns getFirst()
5. DRY parseAndThrowODataError — RESOLVED: moved to AbstractDataverseTask as protected static
6. DRY parseAndThrowError — RESOLVED: moved to AbstractBusinessCentralTask as protected static
7. index.yaml — RESOLVED: description and body reference Dynamics 365, Dataverse, Business Central
8. Nits — RESOLVED

---

### Kestra Guidelines — 2 findings

[REQUEST CHANGES] Production visibility widened for test convenience
Location: AbstractDynamics365Task#getAccessToken (line 80)
Problem: 'protected' was widened to 'public' solely so Mockito doReturn calls from test classes in sub-packages compile. This directly violates the guideline: 'Do not write production code specifically to satisfy tests.' It exposes an internal auth flow as public API on every task class.
Fix: Revert to protected. Add a test-only abstract base in src/test that overrides getAccessToken to return a fixed token. Each test class extends that base instead of relying on Mockito spy.

[Suggestion] apiEndpoint added to AbstractBusinessCentralTask primarily for test redirection
Location: AbstractBusinessCentralTask (line 44)
Problem: Sovereign-cloud justification is plausible, but the scope(RunContext) signature change and propagation to 6 BC task callers was driven by test WireMock redirection. Dataverse achieves the same cleanly through the existing orgUrl property. The description text 'for testing' confirms it. Property is hidden=true which limits UI blast radius.
Fix: If sovereign-cloud is a real requirement, keep but remove 'or for testing' from the description. Otherwise remove it and redirect tests via a test-only subclass overriding baseUrl().

---

### New Findings from This Commit

[Nit] Dead throw after parseAndThrow repeated 9 times in all task run() methods. Declare methods as returning RuntimeException so callers can write 'throw parseAndThrowError(...)' and remove the dummy unreachable statement.

[Nit] ObjectMapper declared in concrete classes despite abstract classes also having one. AbstractBusinessCentralTask and AbstractDataverseTask each declare a private static MAPPER. 8 concrete classes still declare their own. Promoting to protected static final in the abstract class removes all duplicates.

---

### Business Requirements — met
All 9 task classes are implemented with documentation, property groups, examples, WireMock-backed functional tests, and correct error handling.

### Security (OWASP Top 10) — 0 new findings
Auth token not logged. No secrets in outputs. OData filter URL-encoded. package-info.java present for both subpackages. Metadata files present for both subpackages.

---

Verdict: REQUEST CHANGES
The production-for-tests violation on getAccessToken (protected to public) is a non-negotiable guideline point. Fix is small: revert visibility and add a test-only base class in src/test. All prior BLOCK items are cleanly resolved.

**Comment by fdelbrayelle** [comment]
## QA Report — PR #62 (Dynamics 365 Plugin)

**Edition:** Kestra EE (develop image)
**Result:** PASS (all tasks recognized; failures are expected — no Azure credentials configured)

### Summary

| # | Flow | Task Type | Status | Failure reason |
|---|------|-----------|--------|----------------|
| 1 | dataverse_query_accounts | dynamics365.dataverse.Query | FAILED (expected) | Cannot find secret for key AZURE_TENANT_ID |
| 2 | bc_list_companies | dynamics365.businesscentral.ListCompanies | FAILED (expected) | Cannot find secret for key AZURE_TENANT_ID |
| 3 | oneshare_list_nonregression | oneshare.List (non-regression) | FAILED (expected) | Cannot find secret for key TENANT_ID |

### QA Verdict

PASS - Plugin JAR loads, both sub-groups registered, 9 new task types recognized, execution fails correctly at secret resolution (PebbleException), not a plugin crash. Non-regression test shows existing oneshare tasks unaffected.

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/AbstractDataverseTask.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Returns the OAuth2 scope for Dataverse: `{orgUrl}/.default`.
+     */
+    protected String scope(RunContext runContext) throws IllegalVariableEvaluationException {
+        return resolvedOrgUrl(runContext) + "/.default";
+    }
+
+    protected static final ObjectMapper MAPPER = JacksonMapper.ofJson();
+
+    protected static RuntimeException parseAndThrowODataError(int statusCode, String body) {
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/AbstractDataverseTask.java]: Similar to `parseAndThrowError`, extract it to `AbstractDynamics365Task`

File: src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Delete.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        var rRecordId = runContext.render(recordId).as(String.class).orElseThrow();
+
+        var rScope = scope(runContext);
+        var token = getAccessToken(runContext, rScope);
+        var url = baseUrl(runContext) + rEntitySetName + "(" + rRecordId + ")";
+
+        try (var client = new HttpClient(runContext, httpConfiguration())) {
+            var request = HttpRequest.builder()
+                .addHeader("Authorization", "Bearer " + token)
+                .addHeader("OData-MaxVersion", "4.0")
```
Comment (Malaydewangan09) [inline src/main/java/io/kestra/plugin/microsoft365/dynamics365/dataverse/Delete.java]: Extract these magic numbers to constants

File: src/main/resources/metadata/businesscentral.yaml
Comment (Malaydewangan09) [inline src/main/resources/metadata/businesscentral.yaml:1]: File name is incorrect, should be `<subpackage>.yml`

File: src/main/resources/metadata/dataverse.yaml
Comment (Malaydewangan09) [inline src/main/resources/metadata/dataverse.yaml:1]: File name is incorrect, should be `<subpackage>.yml`

**Review by Malaydewangan09 — APPROVED** [review]

**Review by Malaydewangan09 — APPROVED** [review]
